### ISSUE TITLE
Redundant include in implementation file not suggested for removal

### ISSUE THREAD

**Issue by geimer** [include-what-you-use/include-what-you-use#987]
Minimal example:
```cpp
///// foo.h
#include <cstdint>
void foo(std::uint8_t arg);

///// foo.cpp
#include "foo.h"
#include <cstdint>
void foo(std::uint8_t arg) { }
```
Here, `foo.h` is correctly identified as the associated header of `foo.cpp`, and the include of `cstdint` is marked as OK because of using `std::uint8_t` in the function declaration.  However, instead of identifying the second include of `cstdint` in `foo.cpp` as redundant and suggesting it for removal (IMO the expected behavior), IWYU is happy with `foo.cpp`.  In fact, IWYU thinks that it is OK regardless of whether `cstdint` is included or not.

Note that this issue is unrelated to any mappings for `cstdint`, as the same behavior is seen with user-defined types and project headers instead of `std::uint8_t`/`cstdint`.

**Comment by kimgr** [comment]
Does a higher verbosity level give any insights as to its reasoning?

**Comment by geimer** [comment]
As far as I can see, not really.  Please find the full log with `verbose=10` [here](https://gist.github.com/geimer/c548008fdf47a8ac60caaab6517a11a1).  It looks almost the same when commenting out the redundant include of `cstdint` in `foo.cpp` except for some missing lines w.r.t. "found/added include" and "foo.cpp intends-to-provide cstdint and transitive includes". But the "calculating violations" parts at the end of the log are absolutely identical.

**Comment by kimgr** [comment]
I wonder if it has something to do with the fact that the use in `foo.h` is a fwd-decl use, and the one in `foo.cpp` is a full use: https://gist.github.com/geimer/c548008fdf47a8ac60caaab6517a11a1#file-log_verbose10-L821.

A forward-decl is enough for a parameter in a function declaration, but the language requires a full declaration of parameters for function definitions. But with that line of reasoning, IWYU should suggest that `<cstdint>` be replaced by a forward-decl in `foo.h`... It's a nice small repro, so it should be possible to isolate where IWYU missteps in the analysis. Feels like it's confusing itself.

**Comment by kimgr** [comment]
I spent some time today poking around with this. I think you're hitting this check: https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_output.cc#L1782.

I wonder what the effects would be if we brutally removed all `#include`s already present in the associated header. Off the top of my head, that seems like low-hanging fruit for a great reduction in includes, but I'm not sure if I'm missing something.

**Comment by geimer** [comment]
Yes, the comment suggests that this is the place where the extra `#include` is being ignored.  Maybe a line of debug output would make sense here?

Assuming that `bar.h` is already identified as being required for the associated header `foo.h`, I also can't think of any cases where removing the extra `#include` of `bar.h` from `foo.cc` could break things -- except that it may violate policy.  For example, the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Include_What_You_Use) seems to require the extra `#include`.  However, this isn't enforced by the currently implemented behavior either.  To me this feels like an options is required to support different policies.  (Skimming over other issues I got the impression that you're not really in favor of adding more options, but I guess sometimes it's unavoidable...)

**Comment by kimgr** [comment]
Nice, I didn't remember that part of the Google C++ guide:
> Do not rely on transitive inclusions. This allows people to remove no-longer-needed
> #include statements from their headers without breaking clients. This also applies to
> related headers - foo.cc should include bar.h if it uses a symbol from it even if foo.h includes bar.h.

I agree with the sentiment here, but I think it's probably too strict. It's really nice if I can remove `bar.h` from `foo.h`, and none of the `foo` component's clients are affected. But for `foo.cc` to be affected seems like less of a problem. The redundancy feels worse to me.

I experimented with loosening that check, but it caused other problems that needed more context to resolve, so it takes more work to do something about it. I have some ideas, I might follow up on them later.
